dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/CDIContainerImpl.java

```diff
@@ -215,39 +215,41 @@ public void endInitialization(WebSphereCDIDeployment webSphereCDIDeployment) thr
     public void applicationStarted(Application application) throws CDIException {
         CDIContainerEventManager eventManager = cdiRuntime.getCDIContainerEventManager();
         if (eventManager != null) {
-            WebSphereCDIDeployment deployment = getDeployment(application);
-            if (deployment != null) {
-                BeanDeploymentModules modules = deployment.getServices().get(BeanDeploymentModules.class);
-                if (modules != null) {
-                    for (BeanDeploymentModule module : modules) {
-                        if (!module.isWebModule()) {
-                            String id = module.getId();
-                            WebSphereBeanDeploymentArchive bda = deployment.getBeanDeploymentArchive(id);
-                            if (bda != null) {
-                                try (ContextBeginnerEnder contextBeginnerEnder = cdiRuntime.createContextBeginnerEnder().extractComponentMetaData(bda.getArchive())
-                                                                                           .extractTCCL(application).beginContext()) {
-                                    eventManager.fireStartupEvent(module);
-                                }
-                            } else {
-                                throw new IllegalStateException(Tr.formatMessage(tc, "no.bda.for.module.CWOWB1019E", module, module.getId()));
-                            }
-                        }
-                    }
-                }
-            }
+            runForEachModuleWithContext(eventManager::fireStartupEvent, application);
         }
     }
 
     public void applicationStopping(Application application) throws CDIException {
         CDIContainerEventManager eventManager = cdiRuntime.getCDIContainerEventManager();
         if (eventManager != null) {
-            WebSphereCDIDeployment deployment = getDeployment(application);
-            if (deployment != null) {
-                BeanDeploymentModules modules = deployment.getServices().get(BeanDeploymentModules.class);
-                if (modules != null) {
-                    for (BeanDeploymentModule module : modules) {
-                        if (!module.isWebModule()) {
-                            eventManager.fireShutdownEvent(module);
+            runForEachModuleWithContext(eventManager::fireShutdownEvent, application);
+        }
+    }
+
+    /**
+     * This method loops through each module (that is not a WebModule) in an application. For each one it sets
+     * the component metadata associated with that module as well as the TCCL associated with the application
+     * then it runs an action. The action is likely an event form CDIContainerEventManager
+     *
+     * @param action The action to run
+     * @param application the application to run actions
+     */
+    private void runForEachModuleWithContext(Consumer<BeanDeploymentModule> action, Application application) throws CDIException {
```

Comment:
I'd forgotten that it only runs for non-web modules.

I would suggest renaming to `runForEachNonWebModuleWithContext`, or restructuring the code into more easily-named parts.

We should also add a comment to indicate why we want to only run for non-web modules in the existing locations that this is called so that it's easier to track down any problems in this area in future. I'm guessing that the integration with the web container calls these events for web modules for some reason, but we'd need to check.

Reply:
Since this is a private method and there's no likelihood we'll need to do something similar for a different combination of modules, I think renaming is the way to go.

Regarding the comment. `!isWebModule` originates in this commit, so I'll ask Tom if he remembers why he did it this way.